### ISSUE TITLE
write_deflated_file is not flexible enough

### ISSUE THREAD

**Issue by ananthakumaran** [WeTransfer/zip_tricks#21]
Thanks for the library.

I am trying to use zip_tricks instead of the default zip implementation that comes with https://github.com/felixbuenemann/xlsxtream
But the `write_deflated_file` is not flexible enough. I am currently patching the class to solve the issue

```ruby
require 'zip_tricks/streamer'
class ZipTricks::Streamer
  def open_deflated_file(filename)
    add_file_and_write_local_header(filename: filename,
                                    storage_mode: DEFLATED,
                                    use_data_descriptor: true,
                                    crc32: 0,
                                    compressed_size: 0,
                                    uncompressed_size: 0)

    DeflatedWriter.new(@out)
  end

  def close_deflated_file(writable)
    crc, comp, uncomp = writable.finish

    # Save the information into the entry for when the time comes to write
    # out the central directory
    last_entry = @files[-1]
    last_entry.crc32 = crc
    last_entry.compressed_size = comp
    last_entry.uncompressed_size = uncomp
    e = @files.fetch(-1)
    @writer.write_data_descriptor(io: @out, crc32: e.crc32, compressed_size: e.compressed_size, uncompressed_size: e.uncompressed_size)
  end
end

class ZipTricksIO
  def initialize(body)
    @streamer = ZipTricks::Streamer.new(body)
    @writable = nil
    @writer = nil
  end

  def <<(data)
    @writable << data
  end

  def add_file(path)
    close_current_file
    @writer = @streamer.open_deflated_file(path)
    @writable = ZipTricks::Streamer::Writable.new(@writer)
  end

  def close
    close_current_file
    @streamer.close
  end

  private
  def close_current_file
    if @writable
      @streamer.close_deflated_file(@writer)
    end
    @writable = nil
    @writer = nil
  end
end
```

Would you consider adding open_deflated_file/close_deflated_file method in streamer? I can open a PR if you are OK and we can discuss the details in the PR.

**Comment by julik** [comment]
Interesting use case, very fitting what we had to do with Rubyzip ourselves back in the day ("plugging this thing into that other thing"). To be honest, I am not that keen on adding more options to `write_deflated_file` because it is pretty complex as it is already. There is also an inherent issue with things like the zip streamer in that they need to maintain a certain order of operations. Yes, you can expose their internals to some degree, but IMO the more of the sequencing steps you can get wrong, the more they will be gotten wrong. What you are dealing with here is a situation where you need to suspend the execution of a block mid-flight and be able to resume it later. Why not use a Fiber given that it's made for exactly that sort of thing? This little test seems to give me the functionality you are trying to implement:

```
require 'zip_tricks'
require 'set'

class ZipTricksIO
  def initialize(body)
    @streamer = ZipTricks::Streamer.new(body)
    @wf = nil
  end

  def <<(data)
    @wf.resume(:write, data)
  end

  def add_file(path)
    @wf.resume(:close, nil) if @wf
    @wf = Fiber.new do | step, data |
      @streamer.write_deflated_file(path) do |write_sink|
        if step == :write
          write_sink << data
          Fiber.yield
        end
      end
    end
  end

  def close
    @wf.resume(:close, nil) if @wf
    @streamer.close
  end
end

io = ZipTricksIO.new($stdout)
io.add_file('somefile.txt')
io << 'Hello, I am a deflated string'
io.add_file('anotherfile.txt')
io << 'Hello, I am another deflated string'
io.close
```

Since you need this as a narrow case adapter for a specific other library this seems like an acceptable tradeoff to me. It is of course dependent on 2 things: you should not be transporting this IO thing between threads and you are dependent on the XLS writer library to call `close`, otherwise you will be leaking zlib structures along the way.

Would this solve your problem?

**Comment by ananthakumaran** [comment]
I had a mild inclination that this problem could be solved by using Fiber/Thread, but there are multiple reasons for why I didn't pursue it further

1) We don't control the entire stack, we are using Rails/Apache. As such we have to play by it's rule. For streaming response body, rails expectes a enumerator.

```ruby
self.response_body = Enumerator.new do |body|
  xlsx = Xlsxtream::Workbook.new(body, io_wrapper: ZipTricksIO)
  xlsx.write_worksheet('Sheet 1', :auto_format => true) do |sheet|
    records_batch.each do |records|
      records.each do |r|
        sheet << r
      end
    end
  end
  xlsx.close
end
```

In one of my past projects Rails/Thin/EventMachine/EMMysql, we ran into random/hard to debug errors related to the usage of fiber. Basically using Fiber means, I have to make sure Rails & the web server Apache/Thin/Puma don't pass the enumerator across threads. It's probably not worth the effort for me at this time.


2) I was also planning to open a PR in xlsxtream for adding IO::ZipTricks. Using Fiber/thread would reduce the possibility of it getting merged.


As for the API, ruby's File.open supports both use case. If you pass a block, the file is closed automatically. Otherwise it's the responsibility of the user to close it properly.

From a maintainer perspective I understand that adding a new method will increase the possible permutation of call sequence, and it's preferable to keep the permutation to minimum such that user will not be able to do bad things. But sometimes it makes the API less flexible as I have shown.

Again thanks for the library. It wouldn't have been possible for me to stream(or do it in such a short time) if your library is not available. I have presented my arguments and it's upto you to decide.

**Comment by ananthakumaran** [comment]
I will retract my first point. Looks like Enumerator itself is implemented using Fiber and can't be shared across threads, so it's safe to assume Rails won't share it across threads.

**Comment by julik** [comment]
A little correction - here is what you need to support multiple successive writes:

```ruby
class ZipTricksIO
  def initialize(body)
    @streamer = ZipTricks::Streamer.new(body)
  end

  def <<(data)
    @wf.resume(data)
    self
  end

  def add_file(path)
    @wf.resume(:__close__) if @wf

    @wf = Fiber.new do | bytes_or_close_sym |
      @streamer.write_deflated_file(path) do |write_sink|
        loop do
          break if bytes_or_close_sym == :__close__
          write_sink << bytes_or_close_sym
          bytes_or_close_sym = Fiber.yield
        end
      end
    end
  end

  def close
    @wf.resume(:__close__) if @wf
    @streamer.close
  end
end
```

This makes it 30 lines total to provide the adapter that you need. Now to other points that you have

> Basically using Fiber means, I have to make sure Rails & the web server Apache/Thin/Puma don't pass the enumerator across threads.

Formally speaking - yes, you have to. However, I am yet to see a Ruby app server that will take something you iterate over in a response, and pass it between threads. As a matter of fact if a server does use threads at all it is going to constrain most of the things related to a single requesting client to one thread. Thin does not thread at all.

It is tempting to say that

>In one of my past projects Rails/Thin/EventMachine/EMMysql, we ran into random/hard to debug errors related to the usage of fiber

which is our instinct as being engineers ("I've been burned by it once") but can you test the implementation I proposed before saying "it doesn't work"? The only thing I can see happening with that is leaking a resource if the response is not read in full, but this is going to be the case with a deferred writable object as well.

That being said, one of the possible approaches is indeed expanding the imperative API  like this:

```ruby
s = ZipTricks::Streamer.new(out)
s.add_compressed_file_entry(filename: ‘abcedf.txt’, uses_data_descriptor: true)
s << data # Note that here the semantics of this write change a little
s.close_data_descriptor(compressed_size: ..., uncompressed_size: ..., crc32: ...)
s.close
```

That to prevent exposing the `Entry` objects within the Streamer since they are strictly speaking a bit of a private API - they don't change often but they are not necessarily very convenient to use.

Do however compare the complexities. To do it we need to add 1 new method to Streamer. We also need to add a argument to both methods that add the local file header, and we need to branch based on that argument in a substantial way. This adds both complexity and tests, a lot more than what you need to implement the solution I have outlined. _Or_ we need to let the `write_deflated_file` (and the stored sibling) return some kind of sink that you manually close, and trigger that path when you do not provide a block.

```ruby
writer_for_file = write_deflated_file('somefile.txt')
writer_for_file << ...
writer_for_file.close
```

This approach would have my preference, but again - it adds complexity. This does introduce more possible invalid states, so probably I have to bring back the state machine that was removed in f1801fd42c838e78f440a05187bb924e2207cd5c. It does have a downside though that we can no longer return an offset-after-write from `write_..._file`.

WRT including a ZipTricks adapter in the XLS streaming library - I would love to see that happen. I wonder however what @felixbuenemann thinks about it given that his library does not require Ruby 2.0 from what I can see, and we do (by way of kwargs).

If you are absolutely set on not using Fiber I think what you are doing in your description is actually fine - you are using some private methods and whatnot, but if you have tests you will be able to see if anything breaks. Only I would do it at least with a subclass. Another option that you could consider is buffering the the contents of a "file within the XLS" into a Tempfile or something, and pass it through the deflater when you close it. This will be way less "streaming" in the sense that you will be flushing every now and then

**Comment by felixbuenemann** [comment]
I have though about switching to zip_tricks for xlsxtream, so I'm open to PRs.

There have been a number of issue from people looking for true streaming and it has not been possible with rubyzip, because it cannot write the incomplete zip entry headers required to achieve that.

I have no trouble in raising the minimum ruby version requirement if there is a good reason to do so.

Instead of raising the ruby version, xlsxtream could also drop the hard dependency on either rubyzip or zip_tricks and try to detect which gem is available and use the right IO adapter during require.

**Comment by julik** [comment]
@felixbuenemann thanks for your perspective. Detecting at runtime is an option - but you then need to ensure that the ZT version loaded is the one your library supports, at runtime (since you will be depending on whichever one someone added to their Gemfile). I do think that providing a good default is a better way though, because you then reduce the number of choices your users have to make _and_ you remove a setup that you need to maintain with Rubyzip. Given that some Ruby versions will be dropped though I think a major version bump will be necessary. I've filed you a PR using the fiber approach, let's see if that works out. You can then also dispence with the `reopen` call you had to introduce.

@toots I see you did something similar in https://github.com/lidcore/zip_tricks/commit/3acfb0db61abbfbc054344cb059c8085d87c69d7 - is this for the same use case?

**Comment by toots** [comment]
Ha! Thanks for noticing @julik I did it and forgot to report.

We were building a streamed zipped using the evented EventMachine API. In this context, I couldn't find a solution to write a file asynchronously so I added this method to have a callback to execute when the file's data was done being sent.

Couldn't find any other ways to do it with the existing API although it seemed pretty close at times but perhaps I missed something?

**Comment by julik** [comment]
Hehe - my pleasure. Short of using a Fiber like I suggested above - no, not really, there isn't a blessed way to do it. In #24 you could do this as follows:

```
w = zip.write_deflated_file(filename)
aw = AsyncWritable.new(w)
aw.on_done { w.close }
```

That part I am not going to include in the library because I do not condone EventMachine as a phenomenon :-) but for you it will then be a wrapper of about 5 lines. You would also need to ensure you are not writing to multiple Writables criss-cross - since Streamer is not going to suspend any writes waiting for callbacks. Would that work for you?

**Comment by julik** [comment]
@ananthakumaran @felixbuenemann with what's in #24 the implementation for xslxtream specifically becomes this:

```ruby
class ZipTricksIO
  def initialize(body)
    @streamer = ZipTricks::Streamer.new(body)
  end

  def <<(data)
    @wf << data
    self
  end

  def add_file(path)
    @wf.close if @wf
    @wf = @streamer.write_deflated_file(path)
  end

  def close
    @wf.close if @wf
    @streamer.close
  end
end
```

**Comment by felixbuenemann** [comment]
> @felixbuenemann thanks for your perspective. Detecting at runtime is an option […]. I do think that providing a good default is a better way though, […]

You are right, supporting two zip libraries isn't worth the trouble.

> I've filed you a PR using the fiber approach, let's see if that works out.

Thank you. Are you going to release a zip_tricks version with the changes from #24 soon? If so I'd rather wait for that before switching to zip_tricks, because it makes the wrapper much more readable.

> You can then also dispence with the `reopen` call you had to introduce.

This has already been fixed in xlsxtream 1.0.0.

**Comment by toots** [comment]
@julik Let's say that I'm hmm agnostic toward the whole EM thing. Just needed to use it :-)

That looks great to me! Of course, you still need to write files oner after the other, although implementing an asynchronous mutex is kinda interesting:
```
    # This is very naive!
    class Lock
      def initialize
        @pending = []
        @working = false
      end

      def work
        return if @pending.empty?
        @working = true
        @pending.shift.call
      end

      def acquire &block
        @pending << block
        work unless @working
      end

      def release
        @working = false
        work
      end
    end
```

**Comment by julik** [comment]
@toots @felixbuenemann @ananthakumaran Thanks for thinking with me! I've pushed 4.5.0 which includes the deferred writes functionality.

**Comment by julik** [comment]
Closed via #24 and https://github.com/felixbuenemann/xlsxtream/pull/16

**Comment by toots** [comment]
Awesome thanks I'll give it a try asap.
